runner/internal/recorder/recorder.go

```diff
@@ -183,19 +190,9 @@ func (r *Recorder) tickProgress() {
 	}
 }
 
-// updateProgress calls r.config.OnProgress with a new computed Progress
+// updateProgress calls r.onProgress with a new computed Progress
 func (r *Recorder) updateProgress() {
-	r.safeOnUpdateProgress()(r.Progress())
-}
-
-func (r *Recorder) safeOnUpdateProgress() func(Progress) {
-	r.mu.RLock()
-	defer r.mu.RUnlock()
-	onProgress := r.config.OnProgress
-	if onProgress == nil {
-		return func(Progress) {}
-	}
-	return onProgress
```

Comment:
What was the sync lock purpose? The function simply read from readonly (hopefully) fields of the requester.

It seems overkill to me indeed. Is there anything else I missed ?

Reply:
Here it works equally without the locks here, they probably remained from a former implementation that needed them.
However it's more tricky to spot than it looks, for instance:

```go
// 1. Works without locks
func (r *Recorder) updateProgress() {
	onProgress := r.config.OnProgress
	if onProgress == nil {
		onProgress = func (Progress) {}
	}
	onProgress(r.Progress())
}

// 2. Data race!
func (r *Recorder) updateProgress() {
	onProgress := r.config.OnProgress
	if onProgress == nil {
		return
	}
	onProgress(r.Progress())
}

// 3. Data race fixed with locks
func (r *Recorder) updateProgress() {
	r.mu.RLock()
	defer r.mu.RUnlock()
	onProgress := r.config.OnProgress
	if onProgress == nil {
		return
	}
	onProgress(r.Progress())
}
```

I probably I first came up with an implementation similar to 2, then added locks, then refactored to 1 ignoring that would fix the data race 🤷

Reply:
What was the data race about? On implementation 2, the call to `r.Progress` is the one with concurrency (and has its own locks i think). The early return should be a noop. And the single read to `r.config.OnProgress` does not cause data races, or there would still be some in 1.

Am I missing something?

Sidenote: this is out of the scope of pr, as the lock as been removed altogether ^^. We may continue the discussion but i will approve the pr anyway.

Reply:
> Am I missing something?

Yes: the same as I am 😛